### ISSUE TITLE
[xml] wrong cdata rule description and examples

### ISSUE THREAD

**Issue by Vampire** [pmd/pmd#1666]
**Rule:** [MistypedCDATASection](https://pmd.github.io/pmd-6.11.0/pmd_rules_xml_errorprone.html#mistypedcdatasection)

**Description:**
The description and example of the cdata xml rule is wrong.
The start token is not `<!CDATA[`, but `<![CDATA[`

**Comment by adangel** [comment]
Indeed, you're right, thanks for reporting!
The description of the rule is wrong and the example should contain a complete well-formed xml sample.

**Comment by KroArtem** [comment]
It's not clear what is supposed to be in example section. This description looks to be not very helpful:

**Example(s):**

``` xml
An extra [ looks like <!CDATA[[]]>, and an extra ] looks like <!CDATA[]]]>.
```

**Comment by adangel** [comment]
For reference, here's the definition of CDATA-section: https://www.w3.org/TR/2006/REC-xml11-20060816/#sec-cdata-sect

> CDATA sections may occur anywhere character data may occur; they are used to escape blocks of 
> text containing characters which would otherwise be recognized as markup. CDATA sections begin 
> with the string "&lt;![CDATA[" and end with the string "]]&gt;"

Let's break it apart:

1. beginning of a CDATA section: It correctly begins with `<![CDATA[`. I *assume* the rule is supposed to detect wrong placement of the first square bracket, e.g. this: `<!CDATA[[` - it looks like the programmer wanted to start a CDATA section, but got it wrong. The result is, that there is no CDATA section in the text data, which causes follow-up errors, which this rule tries to avoid.
So in the example, we can say this:

```
This text <!CDATA[[ character data ]]> looks like a CDATA section but it isn't.
The first square bracket is misplaced. It should be <![CDATA[ character data ]]>.
```

2. Ending of a CDATA section: It correctly ends with `]]>`. I *assume* the rule is supposed to detect redundant square brackets at the end - since 2 or 3 of such brackets are maybe difficult to visually distinguish. E.g. this `]]]>` is probably not, what the programmer wanted. While it correctly closes the CDATA section, it leaves an additional square bracket in the character data which ends up later in the XML document, which causes follow-up errors. For the example, we could say this:

```
This cdata section <![CDATA[[ character data ]]]> is valid, but it contains an
additional square bracket in the end.
It should probably be just <![CDATA[[ character data ]]>.
```

Please note, that especially the ending part is prone to false positives. E.g. `<![CDATA[a[i+1]]]>` could be exactly, what the programmer wanted.

On the other side, I think, when the redundant square bracket is separated by a space or other character, then, we shouldn't raise a violation, because this indicates a deliberate decision: `<![CDATA[ a[i+1] ]]>`. And since it is separated, it is easier to spot, which square bracket belongs to the ending marker, and which one belongs to the content of the CDATA section.

**Comment by adangel** [comment]
It turns out, the example `<!CDATA[[ character data ]]>` is actually invalid XML - at least the SAX parser we use, doesn't like it. The name "!CDATA" is not a valid tag name. So, no need to test for this case.

This leaves us with

1. `<![CDATA[[ character data ]]>` (additional square bracket at the beginning)
2. `<![CDATA[ character data ]]]>` (additional square bracket at the end)
